### PR TITLE
Add Filterable and FilterableWithIndex typeclasses

### PR DESCRIPTION
**Pull request by 0x706b**

### CODE REVIEW

File: packages/core/src/Prelude/Filterable/index.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    Q,
+    W,
+    X,
+    I,
+    S,
+    R,
+    E
+  >(
+    fa: HKT.Kind<F, C, N, K, Q, W, X, I, S, R, E, A>
+  ) => HKT.Kind<F, C, string, K, Q, W, X, I, S, R, E, B>
```
Comment (qlonik) [inline packages/core/src/Prelude/Filterable/index.ts:31]: Shouldn't this `string` be an `N` type?

Comment (mikearnaldi) [reply packages/core/src/Prelude/Filterable/index.ts:31]: Arguable, I do think N would be a better type here too but the concept of filter is to remove stuff, meaning this N would actually be oversized

**Review by mikearnaldi — CHANGES_REQUESTED** [review]

File: packages/core/src/Prelude/Filterable/index.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
+import type { Either } from "../../Classic/Either"
+import type { Option } from "../../Classic/Option"
+import type { Predicate, Refinement } from "../../Function"
+import type { Separated } from "../../Utils"
+import type { Compactable } from "../Compactable"
+import type { Covariant } from "../Covariant"
+import type * as HKT from "../HKT"
+
+export interface Filterable<F extends HKT.URIS, C = HKT.Auto>
```
Comment (mikearnaldi) [inline packages/core/src/Prelude/Filterable/index.ts:9]: Filterable should be a `type` that intersect its childs like `Filterable = Covariant & Filter & FilterMap & Partition`

**Review by mikearnaldi — CHANGES_REQUESTED** [review]

**Comment by 0x706b** [comment]
I believe this should resolve the requested changes. I separated the functions out into their own folders, similar to `Foldable`, and changed the type of the `N` parameter in the return type from `string` back to `N` for the functions that needed it

**Review by mikearnaldi — APPROVED** [review]

**Comment by mikearnaldi** [comment]
Perfect @0x706b thanks a lot!

**Comment by mikearnaldi** [comment]
Reduced dependencies on Covariant & Compactable (not necessary to pack on filterable instance)

**Pull request merged**